test/jdk/sun/security/pkcs11/Provider/Login.java

```diff
@@ -22,20 +22,33 @@
  */
 
 import java.io.*;
-import java.util.*;
+import java.lang.ref.WeakReference;
 import java.security.*;
 import javax.security.auth.callback.*;
 
 import javax.security.auth.Subject;
 import javax.security.auth.login.FailedLoginException;
+import javax.security.auth.login.LoginException;
+
+import sun.security.pkcs11.SunPKCS11;
 
 public class Login extends PKCS11Test {
 
     private static final String KS_TYPE = "PKCS11";
     private static char[] password;
+    private static SunPKCS11 pkcs11Provider = null;
 
     public static void main(String[] args) throws Exception {
         main(new Login(), args);
+
+        pkcs11Provider.logout();
+        WeakReference<SunPKCS11> weakRef = new WeakReference<>(pkcs11Provider);
+        pkcs11Provider = null;
+        for (int i = 0; i > 100; i++) {
+            System.gc();
+            Thread.sleep(100);
+        }
+        System.out.println("Finish: "+ weakRef.refersTo(null));
     }
```

Comment:
What is this particular block of code for? main(Provider) is non-static and may be called more than once, but then you store the passed-in Provider object into a static field? Also, must the static field use type SunPKCS11? It seems AuthProvider is sufficient and most of the "ap->pkcs11Provider" changes can be avoided?